### PR TITLE
Implement Batch Summarization in MessageHistory Class to manage context length under model's token limit

### LINKED ISSUE #4634: Ver 0.4, and still get the error of "This model's maximum context length is 4097 tokens"
### ⚠️ Search for existing issues first ⚠️

- [X] I have searched the existing issues, and there is no existing issue for my problem

### Which Operating System are you using?

Windows

### Which version of Auto-GPT are you using?

Latest Release

### Do you use OpenAI GPT-3 or GPT-4?

GPT-3.5

### Which area covers your issue best?

Commands

### Describe your issue.

AutoGPT crashes in the same scenario as v0.3.1 when working with "large" local files.

### Upload Activity Log Content

_No response_

### Upload Error Log Content

Traceback (most recent call last):   File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main     return _run_code(code, main_globals, None,   File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code     exec(code, run_globals)   File "/workspace/Auto-GPT/autogpt/__main__.py", line 5, in <module>     autogpt.cli.main()   File "/home/vscode/.local/lib/python3.10/site-packages/click/core.py", line 1130, in __call__     return self.main(*args, **kwargs)   File "/home/vscode/.local/lib/python3.10/site-packages/click/core.py", line 1055, in main     rv = self.invoke(ctx)   File "/home/vscode/.local/lib/python3.10/site-packages/click/core.py", line 1635, in invoke     rv = super().invoke(ctx)   File "/home/vscode/.local/lib/python3.10/site-packages/click/core.py", line 1404, in invoke     return ctx.invoke(self.callback, **ctx.params)   File "/home/vscode/.local/lib/python3.10/site-packages/click/core.py", line 760, in invoke     return __callback(*args, **kwargs)   File "/home/vscode/.local/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func     return f(get_current_context(), *args, **kwargs)   File "/workspace/Auto-GPT/autogpt/cli.py", line 96, in main     run_auto_gpt(   File "/workspace/Auto-GPT/autogpt/main.py", line 197, in run_auto_gpt     agent.start_interaction_loop()   File "/workspace/Auto-GPT/autogpt/agent/agent.py", line 130, in start_interaction_loop     assistant_reply = chat_with_ai(   File "/workspace/Auto-GPT/autogpt/llm/chat.py", line 112, in chat_with_ai     new_summary_message, trimmed_messages = agent.history.trim_messages(   File "/workspace/Auto-GPT/autogpt/memory/message_history.py", line 79, in trim_messages     new_summary_message = self.update_running_summary(   File "/workspace/Auto-GPT/autogpt/memory/message_history.py", line 194, in update_running_summary     self.summary = create_chat_completion(prompt)   File "/workspace/Auto-GPT/autogpt/llm/utils/__init__.py", line 53, in metered_func     return func(*args, **kwargs)   File "/workspace/Auto-GPT/autogpt/llm/utils/__init__.py", line 87, in _wrapped     return func(*args, **kwargs)   File "/workspace/Auto-GPT/autogpt/llm/utils/__init__.py", line 235, in create_chat_completion     response = api_manager.create_chat_completion(   File "/workspace/Auto-GPT/autogpt/llm/api_manager.py", line 61, in create_chat_completion     response = openai.ChatCompletion.create(   File "/home/vscode/.local/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 25, in create     return super().create(*args, **kwargs)   File "/home/vscode/.local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in create     response, _, api_key = requestor.request(   File "/home/vscode/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 226, in request     resp, got_stream = self._interpret_response(result, stream)   File "/home/vscode/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 619, in _interpret_response     self._interpret_response_line(   File "/home/vscode/.local/lib/python3.10/site-packages/openai/api_requestor.py", line 682, in _interpret_response_line     raise self.handle_error_response( openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 4904 tokens. Please reduce the length of the messages. Press any key to continue...

### PR DESCRIPTION
**Pull request by kinance**
### Background
This pull request introduces a new approach to managing the context length when calling the chat completion API to get a running message of previous events in the MessageHistory class. The improvements include processing new events in batches and summarizing them before appending to the existing summary. These changes aim to address the issue where the context length exceeds the model's maximum token limit (4097 tokens for gpt-3.5-turbo).

### Changes

1. Batch Processing of New Events in **update_running_summary** function: To manage the context length and avoid exceeding the maximum token limit, new events are now processed in batches. Each batch's size is determined by the token length limit of the language model. This ensures that the length of the context stays within the allowable limit, avoiding the associated error.
2. Introduction of **summarize_batch** Function: This function handles the summarization of each batch of new events. It receives the batch of new events and the configuration object, logs the prompt to the current log cycle, obtains a summarized message of combining the previous summary and new events in the current batch using the create_chat_completion function, and then updates the summary. The updated summary is also logged in the current log cycle. This approach can keep the context across batches of events because each summarization includes the previous summary and new events.

The changes in this pull request provide a robust solution to manage context length and ensure the agent's memory and conversational context are effectively handled. It mitigates the error of exceeding the model's maximum context length, thereby enhancing the overall performance of the AI assistant.

This can fix #4634 #4019 #4012 #3979 #3681 issues

### Documentation
Detailed comments are added to the code, along with some future enhancements marked as To-Do.

### Test Plan
Added a new test for MessageHistory class to test the batch running summary code by mocking lot of events in the message history and trimming most of the messages.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes. 
- [x] I have run the following commands against my code to ensure it passes our linters:
    ```shell
    black .
    isort .
    mypy
    autoflake --remove-all-unused-imports --recursive --ignore-init-module-imports --ignore-pass-after-docstring autogpt tests --in-place
    ```

### CODE REVIEW

**Review by merwanehamadi — APPROVED** [review]

**Comment by merwanehamadi** [comment]
@kinance it worked I reproduced the issue and then it was gone with your PR. Thanks !

**Comment by ezacngm** [comment]
Thanks for your work!!! I followed your changes and it did mitigate the issue somehow, the assistant can run longer than before.
However for me the error remains, I still fail to run a single project due to this error. "This model's maximum context length is 4097 tokens. However, your messages resulted in 4289 tokens." Apparently the overlong context length is shorter now. 
By any chance some advice please? I'm guessing if the goal I set is too broad and requires intensive search. My prompt is to write a literature review about the diagnosis of the Bipolar Disorder.

File: autogpt/memory/message_history.py [hunk 1]
```diff
@@ -167,20 +168,49 @@ def update_running_summary(self, new_events: list[Message]) -> Message:
...
-Latest Development:
-"""
-{new_events or "Nothing new happened."}
-"""
-'''
+        Latest Development:
+        """
+        {new_events_batch or "Nothing new happened."}
+        """
+        '''
```
Comment (Pwuts) [inline autogpt/memory/message_history.py:213]: This does look better, but it adds unnecessary whitespace to the prompt.

**Comment by Pwuts** [comment]
@ezacngm please post a log when reporting an issue

**Comment by merwanehamadi** [comment]
> Thanks for your work!!! I followed your changes and it did mitigate the issue somehow, the assistant can run longer than before. However for me the error remains, I still fail to run a single project due to this error. "This model's maximum context length is 4097 tokens. However, your messages resulted in 4289 tokens." Apparently the overlong context length is shorter now. By any chance some advice please? I'm guessing if the goal I set is too broad and requires intensive search. My prompt is to write a literature review about the diagnosis of the Bipolar Disorder.


Send your logs here or on discord by sending me a DM. make sure your run --debug please, the logs will be better.
the logs are in the logs folder, zip it.

File: autogpt/memory/message_history.py (hunk 1 above)
Comment (kinance) [reply autogpt/memory/message_history.py:213]: Thanks. Removed them via this PR #4660.

**Comment by KnowhereFern** [comment]
Tested as well and failed for me with 

openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 8734 tokens. Please reduce the length of the messages.

**Comment by EncryptShawn** [comment]
I get reducing content with summary and chunking to manage size but ultimately, I think we need to implement openai's tiktoken and then auto define max-tokens based on the size of prompts we are sending, just as a backup to ensure we never overload the prompt which in result causes the entire process to fail and die in its tracks.  Of course if we allow a prompt and assistant that exceeds the max tokens for the model, then I guess tiktoken would set the max tokens to 0 and it would fail to produce usable results so seems we need things like better summarization and/or chunking and perhaps prompt/assistant truncate > x size on top of the ultimate fail safe of tiktoken.  From what I have read, it sounds like there also needs to be a handful of tokens left as a buffer even with tiktoken.

This issue is currently a show stopper for us so we are debating implementing tiktoken for autogpt if no one else is working it.

**Comment by Pwuts** [comment]
@EncryptShawn what do you mean by _implementing tiktoken_? We currently import the `tiktoken` package and use it to calculate token lengths.

**Comment by JmMndz** [comment]
Already running ver 0.41, but still getting this: openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 8534 tokens. Please reduce the length of the messages.

Command: docker-compose run --rm auto-gpt --gpt3only --continuous

**Comment by Pwuts** [comment]
Please post a log or a stack trace; we can't derive any useful info from just the error itself.

> Already running ver 0.41
Do you mean the release branch? Because v0.4.1 is not released yet.

**Comment by JmMndz** [comment]
> @JmMndz Please post a log or a stack trace; we can't see what's going on from just the error itself.
> 
> > Already running ver 0.41
> 
> Do you mean the release branch? Because v0.4.1 is not released yet.

Yes, sorry, my bad. No ver 0.41, but 0.40 with changes propoused by kinance and Pwuts in this pull request: "Implement Batch Summarization in MessageHistory Class to manage context length under model's token limit"
-Implement Batch Running Summarization to avoid max token error
-Rename test func
-Remove extra spaces in summarization prompt

I will post a log as soon as I get into the error.

**Comment by JmMndz** [comment]
Here is the log:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/app/autogpt/__main__.py", line 5, in <module>
    autogpt.cli.main()
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1635, in invoke
    rv = super().invoke(ctx)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/app/autogpt/cli.py", line 96, in main
    run_auto_gpt(
  File "/app/autogpt/main.py", line 197, in run_auto_gpt
    agent.start_interaction_loop()
  File "/app/autogpt/agent/agent.py", line 130, in start_interaction_loop
    assistant_reply = chat_with_ai(
  File "/app/autogpt/llm/chat.py", line 112, in chat_with_ai
    new_summary_message, trimmed_messages = agent.history.trim_messages(
  File "/app/autogpt/memory/message_history.py", line 80, in trim_messages
    new_summary_message = self.update_running_summary(
  File "/app/autogpt/memory/message_history.py", line 186, in update_running_summary
    self.summarize_batch(batch, cfg)
  File "/app/autogpt/memory/message_history.py", line 224, in summarize_batch
    self.summary = create_chat_completion(prompt)
  File "/app/autogpt/llm/utils/__init__.py", line 53, in metered_func
    return func(*args, **kwargs)
  File "/app/autogpt/llm/utils/__init__.py", line 87, in _wrapped
    return func(*args, **kwargs)
  File "/app/autogpt/llm/utils/__init__.py", line 235, in create_chat_completion
    response = api_manager.create_chat_completion(
  File "/app/autogpt/llm/api_manager.py", line 61, in create_chat_completion
    response = openai.ChatCompletion.create(
  File "/usr/local/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in create
    response, _, api_key = requestor.request(
  File "/usr/local/lib/python3.10/site-packages/openai/api_requestor.py", line 226, in request
    resp, got_stream = self._interpret_response(result, stream)
  File "/usr/local/lib/python3.10/site-packages/openai/api_requestor.py", line 619, in _interpret_response
    self._interpret_response_line(
  File "/usr/local/lib/python3.10/site-packages/openai/api_requestor.py", line 682, in _interpret_response_line
    raise self.handle_error_response(
openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 4132 tokens. Please reduce the length of the messages.

**Comment by EncryptShawn** [comment]
> 
Yeah I guess TikToken is there, but it seemed it wasnt implemented fully otherwise we shouldnt be sending prompts > max tokens to the language model.  Seems the summary fix that is being pushed now with summary token length verification will likely resolve the issue by identifying it being over limit and summarizing down to size and verifying token length of summary while leaving some room for the response.

**Comment by Pwuts** [comment]
@JmMndz thanks, found the issue (hopefully). Should be fixed with 244393e

**Comment by kinance** [comment]
@JmMndz it could also be related to an oversight fixed by #4670. Please add this PR to your local.
